src/components/abstractComponent.ts

```diff
@@ -0,0 +1,436 @@
+   */
+  width(): number;
+  /**
+   * Gets the height of the Component in pixels.
+   */
+  height(): number;
+  /**
+   * Gets the origin of the Component relative to its parent.
+   *
+   * @return {Point}
+   */
+  origin(): Point;
+  /**
+   * Gets the container holding the visual elements of the Component.
+   *
+   * Will return undefined if the Component has not been anchored.
+   *
+   * @return {IContent} content selection for the Component
+   */
+  content(): d3.Selection<void>;
+  /**
+   * Gets the top-level element of the component
+   */
+  element(): d3.Selection<void>;
+ /**
+   * Tell this component to invalidate any caching. This function should be
+   * called when a CSS change has occurred that could influence the layout
+   * of the Component, such as changing the font size.
+   *
+   */
+  invalidateCache(): void;
+}
+
+export abstract class AbstractComponent<D> implements IComponent<D> {
+  private _parent: ComponentContainer;
+
+  protected _content: D;
+  protected _cssClasses = new Utils.Set<string>();
+  protected _destroyed = false;
+  protected _element: d3.Selection<void>;
+  protected _height: number;
+  protected _isAnchored = false;
+  protected _isSetup = false;
+  protected _onAnchorCallbacks: Utils.CallbackSet<GenericComponentCallback<D>>
+    = new Utils.CallbackSet<GenericComponentCallback<D>>();
+  protected _onDetachCallbacks: Utils.CallbackSet<GenericComponentCallback<D>>
+    = new Utils.CallbackSet<GenericComponentCallback<D>>();
+  protected _origin: Point = { x: 0, y: 0 };
+  protected _resizeHandler: IResizeHandler;
+  protected _width: number;
+  protected _xAlignment: string = "left";
+  protected _yAlignment: string = "top";
+
+  protected static _xAlignToProportion: { [alignment: string]: number } = {
+    "left": 0,
+    "center": 0.5,
+    "right": 1,
+  };
+  protected static _yAlignToProportion: { [alignment: string]: number } = {
+    "top": 0,
+    "center": 0.5,
+    "bottom": 1,
+  };
+
+  abstract anchor(selection: d3.Selection<void>): this;
+
+  public onAnchor(callback: GenericComponentCallback<D>) {
+    if (this._isAnchored) {
+      callback(this);
+    }
+    this._onAnchorCallbacks.add(callback);
+    return this;
+  }
+
+  public offAnchor(callback: GenericComponentCallback<D>) {
+    this._onAnchorCallbacks.delete(callback);
+    return this;
+  }
+
+  public requestedSpace(availableWidth: number, availableHeight: number) {
+    return {
+      minWidth: 0,
+      minHeight: 0,
+    };
+  }
+
+  abstract computeLayout(origin?: Point, availableWidth?: number, availableHeight?: number): this;
+
+  public render() {
+    if (this._isAnchored && this._isSetup && this.width() >= 0 && this.height() >= 0) {
+      RenderController.registerToRender(this);
+    }
+    return this;
+  }
+
+  public onResize(resizeHandler: IResizeHandler) {
+    this._resizeHandler = resizeHandler;
+    return this;
+  }
+
+  abstract renderImmediately(): this;
```

Comment:
Group the abstract methods at the top so that it's clear which parts an extending classes need to implement.